### PR TITLE
Missing binders for java.time.LocalDateTime and java.time.LocalDate t…

### PR DESCRIPTION
**Pull request by magiccrafter**
…ypes (#519)

### CODE REVIEW

**Comment by zhumin8** [comment]
Thanks for catching this and contributing!
Looks like there are some checkstyle errors blocking the tests. You might find the google-java-format tool mentioned [here](https://github.com/GoogleCloudPlatform/cloud-spanner-r2dbc/blob/main/CONTRIBUTING.adoc#style) helpful.

**Comment by magiccrafter** [comment]
It should be ok now 🤞. I've added the requested integration tests for #517 here.

**Comment by magiccrafter** [comment]
@elefeint Is there anything else I can help with for this MR?

off-topic: Out of curiosity, when are you planning to release next? I couldn't find any relevant Milestones, and I've seen that the main branch has quite a lot of changes since the last release (1.1.0).

**Comment by magiccrafter** [comment]
@elefeint Is there anything else I can help with for this MR?

off-topic: Out of curiosity, when are you planning to release next? I couldn't find any relevant Milestones, and I've seen that the main branch has quite a lot of changes since the last release (1.1.0).

**Comment by elefeint** [comment]
The codebase is halfway through switchover to the latest version of the spec, 0.9, and the upcoming 1.0 release of the R2DBC SPI. It's blocked on me right now. You are right that it would be good to do a release soon, but unfortunately I can't promise dates.

File: cloud-spanner-spring-data-r2dbc/src/test/java/com/google/cloud/spanner/r2dbc/springdata/it/SpannerR2dbcDialectIntegrationTest.java [hunk 1]
```diff
@@ -105,41 +100,38 @@ public void initializeTestEnvironment() {
...
-            .rowsUpdated()
-            .block();
+        .sql("DELETE FROM PRESIDENT where NAME is not null")
+        .fetch()
+        .rowsUpdated()
+        .block();
   }
 
   @Test
   void testReadWrite() {
```
Comment (zhumin8) [inline cloud-spanner-spring-data-r2dbc/src/test/java/com/google/cloud/spanner/r2dbc/springdata/it/SpannerR2dbcDialectIntegrationTest.java:110]: Unfortunately we don't run integration tests on external contributions, you will need to setup your own project to test this. 
I tried to run this with the comparison and constructor fixed, and find it fails with error:
`Can't find a binder for type: class java.time.LocalDate; `
Tracked a bit back and found that Spring Data fails to convert `java.time.LocalDate` to `com.google.cloud.Date`. Also possibly related, found this [LocalDateConverterOverride](https://github.com/spring-projects/spring-data-relational/blob/main/spring-data-r2dbc/src/main/java/org/springframework/data/r2dbc/convert/R2dbcConverters.java#L260-L269) in spring-data-r2dbc.

Comment (zhumin8) [inline cloud-spanner-spring-data-r2dbc/src/test/java/com/google/cloud/spanner/r2dbc/springdata/it/SpannerR2dbcDialectIntegrationTest.java:110]: Unfortunately we don't run integration tests on external contributions, you will need to setup your own project to test this. 
I tried to run this with the comparison and constructor fixed, and find it fails with error:
`Can't find a binder for type: class java.time.LocalDate; `
Tracked a bit back and found that Spring Data fails to convert `java.time.LocalDate` to `com.google.cloud.Date`. Also possibly related, found this [LocalDateConverterOverride](https://github.com/spring-projects/spring-data-relational/blob/main/spring-data-r2dbc/src/main/java/org/springframework/data/r2dbc/convert/R2dbcConverters.java#L260-L269) in spring-data-r2dbc.

**Comment by magiccrafter** [comment]
@zhumin8 I have defined the converters outside as custom ones, and it is working like a charm. I expected that it would work by adding them to the SpannerDialect. I will troubleshoot the code a bit to figure out the role of the overwriting converters and adapt the PR.

**Comment by magiccrafter** [comment]
@zhumin8 @elefeint Spring is defining some very ambiguous date/time overrides which take precedence over the ones defined in the `SpannerR2dbcDialect ` class. I've raised a question with the maintainers of that spring data library, no answer so far. 

I see two options to proceed from here:
1. Bare minimum. 
Changes: 
    - Add only the missing google data types to the `getSimpleTypes()`:
```java
@Override
public Collection<? extends Class<?>> getSimpleTypes() {
  return Arrays.asList(
    JsonWrapper.class,
    Timestamp.class,
    ByteArray.class,
    Date.class);
}
 ```
 This bit is mandatory. Without that the `cloud-spanner-spring-data-r2dbc` cannot bind and convert from/to any of the 
 time java types and google custom types. If we choose to go with this option the client has to explicitly define every 
 read/write converter in a `R2dbcCustomConversions` bean. 

2. Complete.
Changes: 
    - All from Option 1.
    - Keep the custom read/write converters and define a `R2dbcCustomConversions` bean in a Spring Boot Auto-configuration bean:
    ```java
     @Bean
     @ConditionalOnMissingBean
     public R2dbcCustomConversions r2dbcCustomConversions() {
         return R2dbcCustomConversions.of(
             new SpannerR2dbcDialect(),
             Arrays.asList(
                 new DateToLocalDateConverter(),
                 new LocalDateToDateConverter(),
                 new LocalDateTimeToTimestampConverter(),
                 new TimestampToLocalDateTimeConverter()
             )
         );
     }
    ```
With this second approach, the `cloud-spanner-spring-data-r2dbc` lib will be usable out of the box without forcing the clients to write a custom converter for very basic date and time types. The bean is optional and allows overriding because of the `@ConditionalOnMissingBean` annotation. I've tested it in a project and it works like a charm.

Please let me know what you think and which direction you prefer. Here the question is more like this: 
Do you want to go with option 2?

**Comment by elefeint** [comment]
Since this repository only has the driver and the dialect, we should not be adding Spring Boot autoconfiguration here. In the future, such convenience autoconfigurations will live in Spring Cloud GCP project, but we don't currently have a starter for Spanner R2DBC there.

For the purposes of this repo, document how a user would do (2) in the README -- that's what we've done for [JSON custom converters](https://github.com/GoogleCloudPlatform/cloud-spanner-r2dbc#json-mapping) in the past.

**Pull request closed without merging**